### PR TITLE
Implement and use morph multi-client

### PR DESCRIPTION
**Pull request by cthulhu-rider**
Relates to #746.

### CODE REVIEW

File: pkg/morph/client/notary.go [hunk 1]
```diff
@@ -213,15 +258,26 @@ func (c *Client) UpdateNeoFSAlphabetList(list keys.PublicKeys) error {
...
 		return c.Invoke(contract, fee, method, args...)
 	}
 
 	return c.notaryInvoke(false, contract, method, args...)
 }
 
-func (c *Client) notaryInvokeAsCommittee(contract util.Uint160, method string, args ...interface{}) error {
-	return c.notaryInvoke(true, contract, method, args...)
+func (c *Client) notaryInvokeAsCommittee(method string, args ...interface{}) error {
+	designate, err := c.GetDesignateHash()
```
Comment (alexvanin) [inline pkg/morph/client/notary.go:275]: Keep contract address in the `notaryInvokeAsCommittee` signature. The only usage of role management (designate) contract is just a case, not general rule.

**Comment by alexvanin** [comment]
TLDR: can we try to distinguish client related errors and NeoFS related errors in iterator? If the error is NeoFS related, then do not iterate over clients and return error immediately.

`morph/client` returns various errors from the neo-go client and it's own as well. `Invoke()` has great example:
```go
	resp, err := c.client.InvokeFunction(contract, method, params, cosigner)
	if err != nil {
		return err // <- native client error
	}

	if resp.State != HaltState {
		return &notHaltStateError{state: resp.State, exception: resp.FaultException} // <- NeoFS related business logic error
	}

	if len(resp.Script) == 0 {
		return errEmptyInvocationScript // <- NeoFS related business logic error
	}
```

As for the business logic errors, we don't need to retry the request in all available neo-go clients, because it is not related to the remote node health or connection. As for neo-go client calls, we expect to have `err == nil` while they are healthy and connection is established. So, I propose to define `NeoRPCError` type (not final name), wrap all neo-go client errors and check `errors.As()` in the iterator. What do you think?

File: pkg/morph/client/notary.go [hunk 2]
```diff
@@ -93,17 +102,27 @@ func (c *Client) EnableNotarySupport(opts ...NotaryOption) error {
 		alphabetSource: cfg.alphabetSource,
 	}
 
-	return nil
-}
+	// update client cache
+	for _, cached := range c.clients {
+		cached.notary = c.sharedNotary
+	}
```
Comment (alexvanin) [inline pkg/morph/client/notary.go:108]: This loop sets `sharedNotary` structure for already created clients. I suppose new clients will miss it?

**Comment by cthulhu-rider** [comment]
> TLDR: can we try to distinguish client related errors and NeoFS related errors in iterator? If the error is NeoFS related, then do not iterate over clients and return error immediately. What do you think?

I think it would be very useful. Does it need to be done in separate PR?

**Comment by alexvanin** [comment]
> I think it would be very useful. Does it need to be done in separate PR?

Either way is fine.

File: pkg/morph/client/notary.go (hunk 2 above)
Comment (cthulhu-rider) [reply pkg/morph/client/notary.go:108]: New clients has same setter of the field in `createForAddress` method. Since `EnableNotarySupport` can be called at any time, I added this loop to ensure all clients work in the same mode. I did it as an alternative of pre-clearing the cache before turning on the notary.

File: pkg/morph/client/notary.go (hunk 1 above)
Comment (cthulhu-rider) [reply pkg/morph/client/notary.go:275]: Ok, I'll return previous signature. I'd still implement the method with the proposed signature since two places of code to the same instructions. What would be the best name for it?

File: pkg/morph/client/client.go [hunk 3]
```diff
@@ -37,8 +49,6 @@ type Client struct {
 
 	gas util.Uint160 // native gas script-hash
 
-	designate util.Uint160 // native designate script-hash
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/morph/client/client.go:40]: Caching is implemented on neo-go client side, so I decided to remove the additional level.

File: pkg/morph/client/client.go [hunk 4]
```diff
@@ -308,20 +406,36 @@ func toStackParameter(value interface{}) (sc.Parameter, error) {
 			result.Value = int64(0)
 		}
 	default:
-		return result, fmt.Errorf("chain/client: unsupported parameter %v", value)
+		return result, neofsError{fmt.Errorf("chain/client: unsupported parameter %v", value)}
```
Comment (carpawell) [inline pkg/morph/client/client.go:409]: is there any logic when you choose b/w `noefsError{}` and `wrapNeoFSError()` or its a mistake?

File: pkg/morph/client/client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexvanin) [reply pkg/morph/client/client.go:40]: In fact, all native contracts cached in the neo-go client, including `gas` above, are cached. If you want to delete additional level, maybe we delete this level for all contracts? But I would prefer to see it in a different PR.

File: pkg/morph/client/notary.go (hunk 2 above)
Comment (alexvanin) [reply pkg/morph/client/notary.go:108]: I see, all new clients created by `createForAddress` will already have shared notary, missed that
```go
	c := &Client{
		singleClient: &singleClient{
			notary:       x.sharedNotary,
		},
	}
```

File: pkg/morph/client/client.go (hunk 4 above)
Comment (cthulhu-rider) [reply pkg/morph/client/client.go]: Mistake, fixed

File: pkg/morph/client/client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/morph/client/client.go:40]: Done

**Review by alexvanin — APPROVED** [review]

**Review by carpawell — APPROVED** [review]
